### ISSUE TITLE
<metric> was collected before with same name and label values

### ISSUE THREAD

**Issue by andrei-dascalu** [ansrivas/fiberprometheus#39]
Hello,

I've deployed the fiberprometheus integration with an application running under a load balancer. The application is running a number of copies at the same time, each individually registered to be scraped by Promtheus.

After a while, all metric endpoints return a 500 error saying the following:

```
3 error(s) occurred:
* collected metric "http_requests_total" { label:<name:"method" value:"GETT" > label:<name:"path" value:"/metricslback/logic" > label:<name:"service" value:"callback_api-callback-795f85d8b4-6mfrb" > label:<name:"status_code" value:"200" > counter:<value:1 > } was collected before with the same name and label values
* collected metric "http_request_duration_seconds" { label:<name:"method" value:"GETT" > label:<name:"path" value:"/metricslback/logic" > label:<name:"service" value:"callback_api-callback-795f85d8b4-6mfrb" > label:<name:"status_code" value:"200" > histogram:<sample_count:1 sample_sum:6.21e-05 bucket:<cumulative_count:1 upper_bound:0.005 > bucket:<cumulative_count:1 upper_bound:0.01 > bucket:<cumulative_count:1 upper_bound:0.025 > bucket:<cumulative_count:1 upper_bound:0.05 > bucket:<cumulative_count:1 upper_bound:0.1 > bucket:<cumulative_count:1 upper_bound:0.25 > bucket:<cumulative_count:1 upper_bound:0.5 > bucket:<cumulative_count:1 upper_bound:1 > bucket:<cumulative_count:1 upper_bound:2.5 > bucket:<cumulative_count:1 upper_bound:5 > bucket:<cumulative_count:1 upper_bound:10 > > } was collected before with the same name and label values
* collected metric "http_requests_in_progress_total" { label:<name:"method" value:"GETT" > label:<name:"path" value:"/metricslback/logic" > label:<name:"service" value:"callback_api-callback-795f85d8b4-6mfrb" > gauge:<value:0 > } was collected before with the same name and label values
```

I am not entirely sure about the issue or what I'm looking at, but it only happens with the default metrics collected through fiberprometheus. I would appreciate any pointers as to what I'm doing wrong.

Thanks!

**Comment by ansrivas** [comment]
Hi, 
Thanks for reporting this.

As you reported that there are multiple instances of this service and all of them are being scraped by prometheus, there is a chance that all of them are reporting the same metrics.
You can try to take advantage of `fiberprometheus.NewWith(servicename, namespace, subsystem )` pattern and ensure that somehow they are all different. For e.g. if you are manually running these services, name then `app1`, `app2` so on or if you are running under some orchestrator try to run them all with unique labels. 

Hope this solves the issue for you.

**Comment by andrei-dascalu** [comment]
Hi,

Thing is, they should be unique as the label "service" (as per service name, initialized with the regular New method) appends the hostname (eg: callback_api-callback-795f85d8b4-6mfrb, where the last bit after the first dash is unique as per Kubernetes pod construction)

I will try changing namespace as well, perhaps that can add an extra variable.

Thanks

**Comment by renatojf** [comment]
I am having the same problem. Did you manage to fix it @andrei-dascalu ?
I added a UNIX timestamp label, and still it complains that my metrics were already collected. I find it hard to believe it would collect the same timestamp so fast :/

**Comment by sunhailin-Leo** [comment]
@ansrivas 
* I was fixed this error with latest version of `fiber` and i was tested this patch on my production environment. 👍

**Comment by ansrivas** [comment]
@sunhailin-Leo thank you for the fix, I will have a look tonight.

**Comment by renatojf** [comment]
I actually did an extensive debug on this and reached no actual conclusion. What I know is that there's a leak inside Fiber itself apparently. I have reported this to their Discord, but not much attention was given to it.

I provided them with the examples and the fix I did, which is similar to the one @sunhailin-Leo posted.

```
method := fmt.Sprintf("%s", c.Method())
path := fmt.Sprintf("%s", c.Path())
```

This is the link where I put some of the findings: https://discord.com/channels/704680098577514527/704682020990746624/810870794611195905
